packet_test.go

```diff
@@ -130,85 +184,193 @@ var unmarshalStringTests = []struct {
+					{"posix-rename@openssh.com", "1"},
+				},
+			},
+			want: []byte{
+				0x0, 0x0, 0x0, 0x26,
+				0x1,
+				0x0, 0x0, 0x0, 0x3,
+				0x0, 0x0, 0x0, 0x18,
+				'p', 'o', 's', 'i', 'x', '-', 'r', 'e', 'n', 'a', 'm', 'e', '@', 'o', 'p', 'e', 'n', 's', 's', 'h', '.', 'c', 'o', 'm',
+				0x0, 0x0, 0x0, 0x1,
+				'1',
+			},
+		},
+		{
+			packet: &sshFxpOpenPacket{
+				ID:     1,
+				Path:   "/foo",
+				Pflags: flags(os.O_RDONLY),
+			},
+			want: []byte{
+				0x0, 0x0, 0x0, 0x15,
+				0x3,
+				0x0, 0x0, 0x0, 0x1,
+				0x0, 0x0, 0x0, 0x4, '/', 'f', 'o', 'o',
+				0x0, 0x0, 0x0, 0x1,
+				0x0, 0x0, 0x0, 0x0,
+			},
+		},
+		{
+			packet: &sshFxpWritePacket{
+				ID:     124,
+				Handle: "foo",
+				Offset: 13,
+				Length: uint32(len("bar")),
+				Data:   []byte("bar"),
+			},
+			want: []byte{
+				0x0, 0x0, 0x0, 0x1b,
+				0x6,
+				0x0, 0x0, 0x0, 0x7c,
+				0x0, 0x0, 0x0, 0x3, 'f', 'o', 'o',
+				0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xd,
+				0x0, 0x0, 0x0, 0x3, 'b', 'a', 'r',
+			},
+		},
+		{
+			packet: &sshFxpSetstatPacket{
+				ID:    31,
+				Path:  "/bar",
+				Flags: sshFileXferAttrUIDGID,
+				Attrs: struct {
+					UID uint32
+					GID uint32
+				}{
+					UID: 1000,
+					GID: 100,
+				},
+			},
+			want: []byte{
+				0x0, 0x0, 0x0, 0x19,
+				0x9,
+				0x0, 0x0, 0x0, 0x1f,
+				0x0, 0x0, 0x0, 0x4, '/', 'b', 'a', 'r',
+				0x0, 0x0, 0x0, 0x2,
+				0x0, 0x0, 0x3, 0xe8,
+				0x0, 0x0, 0x0, 0x64,
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		b := new(bytes.Buffer)
+		sendPacket(b, tt.packet)
+		if got := b.Bytes(); !bytes.Equal(tt.want, got) {
+			t.Errorf("sendPacket(%v): got %x want %x", tt.packet, tt.want, got)
 		}
 	}
 }
 
-func sp(p encoding.BinaryMarshaler) []byte {
-	var w bytes.Buffer
-	sendPacket(&w, p)
-	return w.Bytes()
+func sp(data encoding.BinaryMarshaler) []byte {
+	b := new(bytes.Buffer)
+	sendPacket(b, data)
+	return b.Bytes()
 }
 
-var recvPacketTests = []struct {
-	b    []byte
-	want uint8
-	rest []byte
-}{
-	{sp(&sshFxInitPacket{
-		Version: 3,
-		Extensions: []extensionPair{
-			{"posix-rename@openssh.com", "1"},
+func TestRecvPacket(t *testing.T) {
+	var recvPacketTests = []struct {
+		b []byte
+
+		want    uint8
+		body    []byte
+		wantErr error
+	}{
+		{
+			b: sp(&sshFxInitPacket{
+				Version: 3,
+				Extensions: []extensionPair{
+					{"posix-rename@openssh.com", "1"},
+				},
+			}),
+			want: sshFxpInit,
+			body: []byte{
+				0x0, 0x0, 0x0, 0x3,
+				0x0, 0x0, 0x0, 0x18,
+				'p', 'o', 's', 'i', 'x', '-', 'r', 'e', 'n', 'a', 'm', 'e', '@', 'o', 'p', 'e', 'n', 's', 's', 'h', '.', 'c', 'o', 'm',
+				0x0, 0x0, 0x0, 0x01,
+				'1',
+			},
 		},
-	}), sshFxpInit, []byte{0x0, 0x0, 0x0, 0x3, 0x0, 0x0, 0x0, 0x18, 0x70, 0x6f, 0x73, 0x69, 0x78, 0x2d, 0x72, 0x65, 0x6e, 0x61, 0x6d, 0x65, 0x40, 0x6f, 0x70, 0x65, 0x6e, 0x73, 0x73, 0x68, 0x2e, 0x63, 0x6f, 0x6d, 0x0, 0x0, 0x0, 0x1, 0x31}},
-}
+		{
+			b: []byte{
+				0x0, 0x0, 0x0, 0x0,
+			},
+			wantErr: errShortPacket,
+		},
```

Comment:
Let’s add another test here for `errLongPacket`, we can use `0xff` for all four of the bytes, that should trip an `errLongPacket` for as long as `errLongPacket` has any meaning.